subprojects/core/src/main/java/org/gradle/api/internal/project/LifecycleAwareProject.java

```diff
@@ -0,0 +1,1180 @@
+        return delegate.sync(action);
+    }
+
+    @Override
+    public DetachedResolver newDetachedResolver() {
+        return delegate.newDetachedResolver();
+    }
+
+    @Override
+    public Property<Object> getInternalStatus() {
+        executeAllprojectsAction();
+        return delegate.getInternalStatus();
+    }
+
+    @Override
+    public RoleBasedConfigurationContainerInternal getConfigurations() {
+        executeAllprojectsAction();
+        return delegate.getConfigurations();
+    }
+
+    @Override
+    public void configurations(Closure configureClosure) {
+        executeAllprojectsAction();
+        delegate.configurations(configureClosure);
+    }
+
+    @Override
+    public ArtifactHandler getArtifacts() {
+        executeAllprojectsAction();
+        return delegate.getArtifacts();
+    }
+
+    @Override
+    public void artifacts(Closure configureClosure) {
+        executeAllprojectsAction();
+        delegate.artifacts(configureClosure);
+    }
+
+    @Override
+    public void artifacts(Action<? super ArtifactHandler> configureAction) {
+        executeAllprojectsAction();
+        delegate.artifacts(configureAction);
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public org.gradle.api.plugins.Convention getConvention() {
+        executeAllprojectsAction();
+        return extensibleDynamicObject.getConvention();
+    }
+
+    @Override
+    public int depthCompare(Project otherProject) {
+        return delegate.depthCompare(otherProject);
+    }
+
+    @Override
+    public int getDepth() {
+        return delegate.getDepth();
+    }
+
+    @Override
+    public int compareTo(@Nonnull Project o) {
+        return delegate.compareTo(o);
+    }
+
+    @Override
+    public FileOperations getFileOperations() {
+        return delegate.getFileOperations();
+    }
+
+    @Override
+    public ProcessOperations getProcessOperations() {
+        return delegate.getProcessOperations();
+    }
+
+    @Override
+    public PluginContainer getPlugins() {
+        executeAllprojectsAction();
+        return delegate.getPlugins();
+    }
+
+    @Override
+    public void apply(Closure closure) {
+        executeAllprojectsAction();
+        delegate.apply(closure);
+    }
+
+    @Override
+    public void apply(Action<? super ObjectConfigurationAction> action) {
+        executeAllprojectsAction();
+        delegate.apply(action);
+    }
+
+    @Override
+    public void apply(Map<String, ?> options) {
+        executeAllprojectsAction();
+        delegate.apply(options);
+    }
+
+    @Override
+    public PluginManagerInternal getPluginManager() {
+        executeAllprojectsAction();
+        return delegate.getPluginManager();
+    }
+
+    @Override
+    public ConfigurationTargetIdentifier getConfigurationTargetIdentifier() {
+        return delegate.getConfigurationTargetIdentifier();
+    }
+
+    @Override
+    public DynamicObject getAsDynamicObject() {
+        return extensibleDynamicObject;
```

Comment:
❌ This is not present on the cross-project wrapper, but I think giving out the dynamic object is also mutable access.

Reply:
Doing that would actually means executing `allprojects` callbacks eagerly in each Groovy script, even without any property/method access.

Why `LifecycleAwareProject` does implement `DynamicObjectAware`, while `ProblemReportingProject` doesn't?
For `ProblemReportingProject(LifecycleAwareProject)` wrapping [compatibility](https://github.com/gradle/gradle/blob/master/platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/ProblemReportingCrossProjectModelAccess.kt#L1074).

Reply:
If we return the dynamic object here without executing lifecycle allprojects, the users of the dynamic object will not observe changes to the project (or this object) done in the callbacks, which violates the contract of the new API.

Reply:
They will observe the changes as soon as `getProperty/invokeMethod/setProperty/etc` methods will be called